src/sicks3000.cc

```diff
@@ -310,3 +310,330 @@ unsigned short SickS3000::CreateCRC(uint8_t *Data, ssize_t length)
+  "Check whether the correct I/O module has been used, and replace if necessary."
+};
+
+#define MAX_READ_ATTEMPTS 10
+int SickS3000::CheckOperatingDataBlock(int retry_count)
+{
+  if (serial!=NULL)
+  {
+    char interrupt_byte[1] = {0x41};
+    char get_token[20] = {0x00,0x00,0x41,0x44,0x19,0x00,0x00,0x05,0xFF,0x07,0x19,0x00,0x00,0x05,0xFF,0x07,0x07,0x0F,0x9F,0xD0};
+    char request_block_11[10] = {0x00,0x00,0x45,0x44,0x0B,0x00,0x00,0x56,0xFF,0x07};
+    char release_token[20] = {0x00,0x00,0x41,0x44,0x19,0x00,0x00,0x05,0xFF,0x07,0x19,0x00,0x00,0x05,0xFF,0x07,0x00,0x00,0xE7,0xB8};
+    char rx_data[256] = {0xFF};
+    int read_bytes = 0;
+    int written_bytes = 0;
+
+    //Send the interrupt byte, to temporarily stop continuous data transmission
+    //note: multiple sends to ensure continuous mode is interrupted
+    if(serial->WritePort(interrupt_byte, &written_bytes, 1)) {return -1;}
+    ros::Duration(0.1).sleep();
+    
+    //Acquire system token, so that we may query the S3000's data blocks
+    if(serial->WritePort(get_token, &written_bytes, 20) < 0) {return -1;}
+    if(!serial->BlockOnRead(100)) {return -1;}
+    ros::Duration(0.1).sleep();
+
+    //Reuqest data block 11, which contains the error display data
+    if(serial->WritePort(request_block_11, &written_bytes, 10) < 0) {return -1;}
+    if(!serial->BlockOnRead(100)) {return -1;}
+    ros::Duration(0.1).sleep();
+    
+    //Read in the data block
+    if(serial->ReadPort(rx_data, &read_bytes, 256) < 0) {return -1;}
+    if(rx_data[8] != 0x0B)
+    {
+      if(retry_count < MAX_READ_ATTEMPTS)
+      {
+        //Release the system token
+        if(serial->WritePort(release_token, &written_bytes, 20) < 0) {return -1;}
+        if(!serial->BlockOnRead(100)) {return -1;}
+        ros::Duration(0.1).sleep();
+        CheckOperatingDataBlock(++retry_count);
+      }
+      else
+      {
+        ROS_INFO("Unable to read operating data block!");
+        return -1;
+      }
+    }
+    else
+    {
+      ROS_INFO("Operating data block successfully read on attempt %d", retry_count);
+      //Release the system token
+      if(serial->WritePort(release_token, &written_bytes, 20) < 0) {return -1;}
+      if(!serial->BlockOnRead(100)) {return -1;}
+      
+      uint8_t device_state = static_cast<unsigned char>(rx_data[14]) & 0x03;
+      uint32_t display_data = static_cast<unsigned char>(rx_data[16]) + 
+        (static_cast<unsigned char>(rx_data[17]) << 8*1) + 
+        (static_cast<unsigned char>(rx_data[18]) << 8*2) + 
+        (static_cast<unsigned char>(rx_data[19]) << 8*3);
+      ROS_INFO("device_state %02X display_data %08X led_code %04X", device_state, display_data, (display_data & 0x0000FFFF));
+      
+      for(int i = 0; i < NUM_S3000_ERROR_CODES; i++)
+      {
+        if((display_data & 0x0000FFFF) == display_code[i])
+        {
+          ROS_INFO("%s", s3000_error_causes[i]);
+          ROS_INFO("%s", s3000_error_resolution[i]);
+          break;
+        }
```

Comment:
Anyway, the other important thing here is - where do we want to put this information?  I'm guessing we want this in the gui, so it should go out on a topic right?  Diagnostics?